src/analyzer/protocol/mysql/mysql-protocol.pac

```diff
@@ -240,35 +269,141 @@ type Handshake_Response_Packet = case $context.connection.get_version() of {
+		CLIENT_PLUGIN_AUTH -> auth_plugin: EmptyOrNUL_String;
+		0x0 -> none_2    : empty;
+	};
+	have_attrs  : case ( cap_flags & CLIENT_CONNECT_ATTRS ) of {
+		CLIENT_CONNECT_ATTRS -> conn_attrs: Handshake_Connection_Attributes;
+		0x0 -> none_3    : empty;
+	};
+	have_zstd   : case ( cap_flags & CLIENT_ZSTD_COMPRESSION_ALGORITHM ) of {
+		CLIENT_ZSTD_COMPRESSION_ALGORITHM -> zstd_compression_level: uint8;
+		0x0 -> none_4    : empty;
+	};
+} &let {
+	update_auth_plugin: bool = $context.connection.set_auth_plugin(auth_plugin)
+		&if( cap_flags & CLIENT_PLUGIN_AUTH );
+	update_expected: bool = $context.connection.update_expected_from_auth()
+		&if( cap_flags & CLIENT_PLUGIN_AUTH );
 };
 
 type Handshake_Response_Packet_v10 = record {
 	cap_flags   : uint32;
 	max_pkt_size: uint32;
 	char_set    : uint8;
-	pad         : padding[23];
-	credentials : Handshake_Credentials_v10[] &until($input.length() == 0);
+	pad         : padding[23]; 
+	use_ssl     : case ( cap_flags & CLIENT_SSL ) of {
+		CLIENT_SSL -> none    : empty;
+		default -> plain: Handshake_Plain_v10(cap_flags);
+	};
 } &let {
 	deprecate_eof: bool = $context.connection.set_deprecate_eof(cap_flags & CLIENT_DEPRECATE_EOF);
+	client_query_attrs: bool = $context.connection.set_client_query_attrs(cap_flags & CLIENT_QUERY_ATTRIBUTES);
 };
 
 type Handshake_Response_Packet_v9 = record {
 	cap_flags    : uint16;
 	max_pkt_size : uint24le;
 	username     : NUL_String;
 	auth_response: NUL_String;
-	have_db      : case ( cap_flags & 0x8 ) of {
-		0x8 -> database: NUL_String;
+	have_db      : case ( cap_flags & CLIENT_CONNECT_WITH_DB ) of {
+		CLIENT_CONNECT_WITH_DB -> database: NUL_String;
 		0x0 -> none    : empty;
 	};
 	password     : bytestring &restofdata;
 };
 
+# SHA2 Auth
+
+type SHA2_Auth_Packet = record {
+	state: bytestring &restofdata;
+} &let {
+	update_state           : bool = $context.connection.update_state(COMMAND_PHASE)
+		&if(state[0] == FAST_AUTH_SUCCESS);
+	update_conn_expectation: bool = $context.connection.set_next_conn_expected(EXPECT_PUB_KEY)
+		&if(state[1] == REQUEST_PUBLIC_KEY);
+};
+
+type Public_Key_Packet = record {
+	pad    : uint8;
+	pub_key: bytestring &restofdata;
+} &let {
+	update_expectation: bool = $context.connection.set_next_expected(EXPECT_SHA2_AUTH_RESP);
+};
+
+type SHA2_Auth_Response_Packet = record {
+	data: bytestring &restofdata;
+} &let {
+	update_state: bool = $context.connection.update_state(COMMAND_PHASE);
+};
+
+# Auth Switch
+
+type Auth_Switch_Response_Packet = record {
+	data  : bytestring &restofdata;
+} &let {
+	update_expected: bool = $context.connection.update_expected_from_auth();
+};
+
+# Connection Phase
+
+type Connection_Phase_Packets = case $context.connection.get_conn_expectation() of {
+	EXPECT_HANDSHAKE        -> handshake_resp   : Handshake_Response_Packet;
+	EXPECT_SHA2_AUTH        -> sha2_auth        : SHA2_Auth_Packet;
+	EXPECT_PUB_KEY          -> pub_key          : Public_Key_Packet;
```

Comment:
I'm starting to wonder whether there's a need to distinguish the individual states (which are specific to sha2):

It seems it should be possible to slurp client and server data as a generic "AuthMoreData" messages/events until the server sends and OK or Error packet, then switch into the COMMAND_PHASE.

I've sketched this with a generic `mysql_auth_more_data()` event for the server side and believe this could work.

@AmazingPP - would you want to iterate on some of the feedback? Otherwise it might be easier for me to push some changes I have locally in a new branch.

Reply:
Here's a branch:
https://github.com/zeek/zeek/compare/topic/awelzel/mysql-amazing-pp-issue-2716-additions

The top commit is currently mysql: Add mysql_auth_plugin, mysql_auth_more_data and mysql_auth_switch_request events. It removes the SHA2 specific parsing and just treats the initial messages in an opaque way. This should allow users to work with the auth data in script land a bit if they really want, or forward auth data into a separate analyzer in the future.

Let me know what you think (I also pinged you on Slack ;-) )

Reply:
@AmazingPP - I hope you're feeling better!

I've opened a new PR at #3816 that simplifies this. Also extended the query attribute parsing. I'm going to close this PR in favor of getting #3816 in. Take a look and happy to hear what you think.